### PR TITLE
Multiple images support

### LINKED ISSUE #649: [PeerTube] Implement banners
Added with PeerTube v3.2.0, PeerTube's account and video-channels support banner

I did not find an example URL

### LINKED ISSUE #763: [YouTube] Couldn't get high quality thumbnails
server response
截屏2021-12-13 下午2 25 23

### PR DESCRIPTION
**Pull request by AudricV**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [ ] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
      No, because it requires to do changes in it which I don't know how to do, at least in the best way.
- [ ] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.
      No, because I don't know how to do the app support, at least properly, such as implementing options to select an image which fits the resolution of the UI element on which it will be shown (or a highest one if it is already in cache), selecting maximum or minimum resolutions.

This pull request introduces support of multiple images instead of a single image URL.

This will allow control over the image quality used, and also improve it in several places, especially on YouTube.

These changes of course will benefit to all extractor clients, and not only to NewPipe itself: Piped is a good example on which low quality of images can be easily seen.

<details>
<summary><strong>Click here to hide or show how images of Creative Commons channel on Piped looks and how they would look if best resolutions are used instead</strong></summary>

| Current look | Look if higher images provided are used |
|:------------:|:---------------------------------------:|
|     Current look    |                  Look if higher images provided are used                  |
</details>

## Concept implementation and code changes details

Data of images are handled by a specific new class, `Image`, containing the URL to an image, its height and width, if they are known; otherwise, the relevant constants are returned (`HEIGHT_UNKNOWN` and `WIDTH_UNKNOWN`).

Contrary to a previous attempt (#268), the real size is so returned instead of a quality level. In my opinion, it is not to the extractor to decide if images are high or small, but to clients, as each client has its own needs.

Images are returned as unmodifiable lists, because you should be not able to modify data extracted by extractors in its objects, i.e. insert a new stream in a stream list, clients who want to do so should use copies of the lists or their elements (note: that's the case in some places and this should be fixed, but that's to be discussed in a separated issue).

The images methods (and attributes) of extractor classes have been updated to reflect the changes:

<details>
  <summary><strong>Click here to hide or show image methods (and attributes) changes of extractor classes</strong></summary>

  | Class/Interface | Method(s) removed<br><br>(getters return type: `String`; setters return type: `void`) | Method(s) added<br><br>(getters return type: `List<Image>`; setters return type: `void`) |
  |:---:|:---:|:---:|
  | `InfoItem` | - `getThumbnailUrl()`<br>- `setThumbnailUrl(String)` | - `getThumbnails()`<br>- `setThumbnails(List<Image>)` |
  | `CommentsInfoItem` | - `getUploaderAvatarUrl()`<br>- `setUploaderAvatarUrl(String)` | - `getUploaderAvatars()`<br>- `setUploaderAvatars(List<Image>)` |
  | `StreamInfoItem` | - `getUploaderAvatarUrl()`<br>- `setUploaderAvatarUrl(String)` | - `getUploaderAvatars()`<br>- `setUploaderAvatars(List<Image>)` |
  | `InfoItemExtractor` | - `getThumbnailUrl()` | - `getThumbnails()` |
  | `CommentsInfoItemExtractor` | - `getUploaderAvatarUrl()` | - `getUploaderAvatars()` |
  | `StreamInfoItemExtractor` | - `getUploaderAvatarUrl()` | - `getUploaderAvatars()` |
  | `ChannelExtractor` | - `getAvatarUrl()`<br>- `getBannerUrl()`<br>- `getParentChannelAvatarUrl()` | - `getAvatars()`<br>- `getBanners()`<br>- `getParentChannelAvatars()` |
  | `PlaylistExtractor` | - `getUploaderAvatarUrl()`<br>- `getThumbnailUrl()`<br>- `getBannerUrl()`<br>- `getSubChannelAvatarUrl()` | - `getUploaderAvatars()`<br>- `getThumbnails()`<br>- `getBanners()`<br>- `getSubChannelAvatars()` |
  | `StreamExtractor` | - `getThumbnailUrl()`<br>- `getUploaderAvatarUrl()`<br>- `getSubChannelAvatarUrl()` | - `getThumbnails()`<br>- `getUploaderAvatars()`<br>- `getSubChannelAvatars()` |
  | `ChannelInfo` | - `getParentChannelAvatarUrl()`<br>- `setParentChannelAvatarUrl(String)`<br>- `getAvatarUrl()`<br>- `setAvatarUrl(String)`<br>- `getBannerUrl()`<br>- `setBannerUrl(String)` | - `getParentChannelAvatars()`<br>- `setParentChannelAvatars(List<Image)`<br>- `getAvatars()`<br>- `setAvatars(List<Image>)`<br>- `getBanners()`<br>- `setBanners(List<Image>)` |
  | `PlaylistInfo` | - `getThumbnailUrl()`<br>- `setThumbnailUrl(String)`<br>- `getBannerUrl()`<br>- `setBannerUrl(String)`<br>- `getUploaderAvatarUrl()`<br>- `setUploaderAvatarUrl(String)`<br>- `getSubChannelAvatarUrl()`<br>- `setSubChannelAvatarUrl(String)` | - `getThumbnails()`<br>- `setThumbnails(List<Image>)`<br>- `getBanners()`<br>- `setBanners(List<Image)`<br>- `getUploaderAvatars()`<br>- `setUploaderAvatars(List<Image>)`<br>- `getSubChannelAvatars()`<br>- `setSubChannelAvatars(List<Image>)` |
  | `StreamInfo` | - `getThumbnailUrl()`<br>- `setThumbnailUrl(String)`<br>- `getUploaderAvatarUrl()`<br>- `setUploaderAvatarUrl(String)`<br>- `getSubChannelAvatarUrl()`<br>- `setSubChannelAvatarUrl(String)` | - `getThumbnails()`<br>- `setThumbnails(List<Image>)`<br>- `getUploaderAvatars()`<br>- `setUploaderAvatars(List<Image>)`<br>- `getSubChannelAvatars()`<br>- `setSubChannelAvatars(List<Image>)` |
</details>

## Services implementations

Several methods have been added, changed and/or removed, depending of the services. These changes can be found in the commits of this PR.

The implementation of multiple images support differs between each service.

- `YouTube`: images are got from `thumbnails` arrays returned by YouTube, except for mixes when they are extracted (and so when they do not come from a related item or a search result item), for which image and their resolutions are hardcoded:
    - `default.jpg`: 120x90px;
    - `mqdefault.jpg`: 320x180px;
    - `hqdefault.jpg`: 480x360px.
- `SoundCloud`: resolutions are always hardcoded:
  <details>
    <summary><strong>Click here to hide or show image resolutions used on SoundCloud</strong></summary>

    - for artworks and avatars, image and resolutions returned are:
        - `mini`: 16x16px;
        - `t20x20`: 20x20px;
        - `small`: 32x32px;
        - `badge`: 47x47px;
        - `t50x50`: 50x50px;
        - `t60x60`: 60x60px;
        - `t67x67`: 67x67px;
        - `large`: 100x100px (the resolution provided by SoundCloud in its internal (and public?) API);
        - `t120x120`: 120x120px;
        - `t200x200`: 200x200px;
        - `t240x240`: 240x240px;
        - `t250x250`: 250x250px;
        - `t300x300`: 300x300px;
        - `t500x500`: 500x500px.
    - for visuals/user banners, image and resolutions returned are (the resolution provided by SoundCloud in its internal (and public?) API is `original`):
        - `t1240x260`: 1240x260px;
        - `t2480x520`: 2480x520px.
  </details>
- `PeerTube`: resolution of images are only provided, and so known, for banners and avatars. The implementation uses first `avatars` and `banners` JSON arrays, as they contain multiple images; `avatar` and `banner` objects are also used as a fallback if these arrays are not present or empty (so compatibility with old instances is kept);
- `Bandcamp`: as images may be not squares, only image resolutions which preserve aspect ratios are selected (and hardcoded). As a matter of fact, only one dimension of an image is known per image ID:
  <details>
    <summary><strong>Click here to hide or show Bandcamp image IDs and their dimension known</strong></summary>

    - `10`: 1200px wide;
    - `101`: 90px high;
    - `170`: 422px high;
    - `171`: 646px high;
    - `20`: 1024px wide;
    - `200`: 420px high;
    - `201`: 280px high;
    - `202`: 140px high;
    - `204`: 360px high;
    - `205`: 240px high;
    - `206`: 180px high;
    - `207`: 120px high;
    - `43`: 100px high;
    - `44`: 200px high.
  </details>

  This does not apply on banners, for which only the banner image URL extracted from the website is used, and so its dimensions are not known. See why this has been chosen in the `Known limitations/issues` section above.
- `MediaCCC`: There is only one image for contents. A switch to the higher image resolution has been also made in this PR.

To help with implementations and harcoded image resolutions, which are always suffixes to image URLs and/or paths, a class has been added to manage them: `ImageSuffix`. An object of this class stores an image suffix string to add and the height and width corresponding to the image itself.

## Known limitations/issues

- `Multi-service`:
  - original images are not returned, for performance and size purposes;
  - no changes have been also made on `Frameset`s and on `StreamSegment`s: changes required to implement multi-image support on these classes should be made in a separate PR.
- `YouTube`:
  - videos which have a high resolution thumbnail are advertised to have a full HD size, even if it should be only HD. This is a YouTube issue;
  - some videos can have a high resolution thumbnail and it can be not returned by YouTube. This is a YouTube issue and also happens with the official API;
  - not all resolutions are returned, depending of the type of the content requested (video, short, channel, playlist, mix, search). This is a YouTube limitation, at least from the internal API: only image resolutions which would fit the best in the desktop website image elements are returned;
- `SoundCloud`: if default prefixes are changed by SoundCloud (the ones returned in the image URL provided by the internal API), image extraction would be broken;
- `PeerTube`: resolution of thumbnails is not known. Should an issue be opened in [the PeerTube repository](https://github.com/Chocobozzz/PeerTube) to request the addition of it?
- `Bandcamp`: among the image IDs used for avatars and covers, some of them does not seem to match the same dimension property known on banners, according to the following test made on [the banner URL](https://f4.bcbits.com/img/0012957514_100.png) returned for the artist used in `BandcampChannelExtractorTest`:
  <details>
    <summary><strong>Click here to hide or show results</strong></summary>

    - `10`: maximum size provided (975x180px);
    - `101`: height respected (90px);
    - `170`: height not respected and maximum size not provided (750x138px);
    - `171`: maximum size provided (975x180px);
    - `20`: maximum size provided (975x180px);
    - `200`: maximum size provided (975x180px);
    - `201`: height not respected and maximum size not provided (750x138px);
    - `202`: height not respected and maximum size not provided (375x69px);
    - `204`: height not respected and maximum size not provided (960x177px);
    - `205`: height not respected and maximum size not provided (640x118px);
    - `206`: height not respected and different resolution provided (480x89px);
    - `207`: height not respected and different resolution provided (320x59px);
    - `43`: height respected (542x100px);
    - `44`: maximum size provided (975x180px).
  </details>

  This should be investigated later and not in this PR.
- `MediaCCC`: image sizes are not provided and so not known. Should an issue be opened in [the MediaCCC repository](https://github.com/voc/voctoweb) to request the addition of it?

## Changes on tests

Changes were also required on the tests structure, and the ones made are the following ones:

<details>
<summary><strong>Click here to hide or show changes on the tests structure</strong></summary>

| Class/Interface | Method(s) removed<br><br>(return type: `void`) | Method(s) added<br><br>(return type: `void`) |
|:---:|:---:|:---:|
| `BaseChannelExtractorTest` | - `testAvatarUrl()`<br>- `testBannerUrl()` | - `testAvatars()`<br>- `testBanners()` |
| `BasePlaylistExtractorTest` | - `testThumbnailUrl()`<br>- `testBannerUrl()`<br>- `testUploaderAvatarUrl()` | - `testThumbnails()`<br>- `testBanners()`<br>- `testUploaderAvatars()` |
| `BaseStreamExtractorTest` | - `testUploaderAvatarUrl()`<br>- `testSubChannelAvatarUrl()`<br>- `testThumbnailUrl()` | - `testUploaderAvatars()`<br>- `testSubChannelAvatars()`<br>- `testThumbnails()` |
</details>

Image test methods which used URL in their name have been renamed to reflect the multiple images support changes.

A default test on image collections has been added and ensures that:
- image collections are not null;
- for each image of the collections, the following conditions are met:
  - its URL is secure;
  - its height and width are greater or equal to the relevant unknown constants.

Each test is responsible to call this default method or use custom ones like it is made on YouTube (using the default test and asserting that each image URL contains the string `yt`) and Bandcamp (using the default test and asserting that each image URL contains the string `f4.bcbits.com/img` and ends with `.jpg` or `.png`).

## Other changes

A few other changes have been also made in this PR:
- licence headers have been moved to the top of the files modified, where it was applicable and it wasn't already the case;
- `YoutubeMusicSearchExtractor`'s `InfoItem`s have been moved to external classes, which do not depend of the YouTube ones anymore (it was useless to do so, as the objects of YouTube are not ordered in the same way/are not the same than on YouTube Music) in order to increase readability and perform code changes easier;
- unneeded `public` modifiers have been removed from edited test classes;
- missing `@Test` annotations have been added on old Junit 4 tests on modified test classes and the corresponding tests have been fixed if needed;
- single class imports are used instead of wildcard imports in test classes modified when it was not the case;
- a little bit of code has been improved/refactored in some places of the files changed with this PR.

Closes #649, fixes #763.

### CODE REVIEW

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java
Comment (Isira-Seneviratne) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java:74]:
```
          return Stream.of(Jsoup.parse(html).getElementById("customHeader"))
                    .filter(Objects::nonNull)
                    .flatMap(element -> element.getElementsByTag("img").stream())
                    .map(element -> element.attr("src"))
                    .filter(url -> !isNullOrEmpty(url))
                    .map(url -> new Image(replaceHttpWithHttps(url), HEIGHT_UNKNOWN, WIDTH_UNKNOWN))
                    .collect(Collectors.collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
```

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java
Comment (Isira-Seneviratne) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java:76]:
```suggestion
          return Stream.of(Jsoup.parse(html).getElementById("customHeader"))
                    .filter(Objects::nonNull)
                    .flatMap(element -> element.getElementsByTag("img").stream())
                    .map(element -> element.attr("src"))
                    .filter(url -> !isNullOrEmpty(url))
                    .map(url -> new Image(replaceHttpWithHttps(url), HEIGHT_UNKNOWN, WIDTH_UNKNOWN))
                    .collect(Collectors.collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
```

Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampChannelExtractor.java:76]: Done, I also applied your suggestion in other places I found it was relevant to do the changes.

**Review by B0pol — COMMENTED** [review]
> Multiservice:
> original images are not returned, for performance and size purposes

I don’t understand why. We provide all resolutions so that the client can choose, so why remove original? If it’s too large the client can decide not to use it.

> PeerTube: resolution of thumbnails is not known. Should an issue be opened in [the PeerTube repository](https://github.com/Chocobozzz/PeerTube) to request the addition of it?

I think yes

> Contrary to a previous attempt (https://github.com/TeamNewPipe/NewPipeExtractor/pull/268), the real size is so returned instead of a quality level. In my opinion, it is not to the extractor to decide if images are high or small, but to clients, as each client has its own needs.

But how will the client decide if both width and height quality are unknown?
Taking example with PeerTube, previewThumbnail is high quality, thumbnailPath is low quality, we return both as unknown, how is the client supposed to know which one to choose?

While if we have LOW, MEDIUM, HIGH (maybe LOWEST, HIGHEST too), we can set previewPath to HIGH and thumbnailPath to low and the client will know even if there is both HEIGHT_UNKNOWN and WIDTH_UNKNOWN

IMO we should both return actual width/height and estimated quality by extractor.
So that client can choose with size if they want, but when it’s unknown they can use what extractor tells (or even always).
Maybe  we should also sort by Lists by image quality, either from lowest to highest or the other way.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudChannelInfoItemExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudChannelInfoItemExtractor.java:32]: It is provided by default, see https://github.com/TeamNewPipe/NewPipeExtractor/pull/889/files#diff-a1883494e6e8d2c60ffe76496ea85b18be989228c7418d51d31f938907842197R365-R366 and https://github.com/TeamNewPipe/NewPipeExtractor/pull/889/files#diff-a1883494e6e8d2c60ffe76496ea85b18be989228c7418d51d31f938907842197R382-R383 for banners

**Comment by AudricV** [comment]
> Maybe we should also sort by Lists by image quality, either from lowest to highest or the other way.

I asked about this in the IRC channel and here is the conversation I had with Stypox ([Matrix link](https://matrix.to/#/!zdymnitarxhLlVGPHV:libera.chat/$50yPhcpHdDvBEkMMmzFfbNc6uVyL5A3rC0K9w3L2EUM?via=libera.chat&via=matrix.org&via=envs.net)):

Me:
> How do you think a list of images should be sorted?
> I want your advice because I am implementing support of extracting multiple images provided by services in the extractor.
> For now, I'm using images' height because it seems to be the most dimension available, even if it can be unknown in some cases (such as in some Bandcamp image formats which are based on the width).
> For reference, here is the code I use in Java to do the sorting of a list:
> The Image class has three properties: URL, height and width
> `imageList.sort(Comparator.comparingInt(Image::getHeight));`

Stypox:
> I don't think they should not be sorted at all
> If somebody wants to sort them, they can do that on their own, we don't need to do unneeded computation
> Just like video streams are not sorted

Me:
> Ok, then I will only make image lists unmodifiable
> But this will be needed in clients such as NewPipe

Stypox:
> Mmmh, I don't think so, in NewPipe we just need a filter that chooses the best fitting thumbnail according to a size criteria (or something like that), which doesn't require any kind of sorting (on a side note, sorting would run in O(nlogn), while without sorting it is O(n))

File: extractor/src/main/java/org/schabi/newpipe/extractor/Image.java
Comment (TobiGr) [inline extractor/src/main/java/org/schabi/newpipe/extractor/Image.java:32]: The URL and its getters/setters should be annotated as either `@Nullable` or `@NonNull` to make spotting potential bugs easier.

File: extractor/src/main/java/org/schabi/newpipe/extractor/InfoItem.java
Comment (TobiGr) [inline extractor/src/main/java/org/schabi/newpipe/extractor/InfoItem.java:65]: Please annotate these methods as either `@Nullable` or `@NonNull`. From what I've seen so far, they are likely `@NonNull` and empty if no images were found.

File: extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItem.java
Comment (TobiGr) [inline extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItem.java:67]: please annotate null-ability.

**Review by TobiGr — COMMENTED** [review]
Full review is going to come in the next days. The code I have read so far, looks good :+1: 

I agree with the comments made by @B0pol 

Just a note: please annotate the new methods which are going to be accessible for library users. I did not highlight every usage, just a few.

File: extractor/src/main/java/org/schabi/newpipe/extractor/stream/StreamInfo.java
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/stream/StreamInfo.java:183]: Why are you removing this?

**Review by Stypox — CHANGES_REQUESTED** [review]
The code structure looks well done, thank you! I reviewed the code but I did not check whether the URLs you extract actually point somewhere, but I trust you and (in part) the tests for that. I also didn't check if there are more thumbnails that can be extracted, I also trust you here (though even if there were, it wouldn't be a big problem). I have just a few comments + B0pol&TobiGr's comments need to be solved, too.

**Comment by Dertyiula** [comment]
Is there development going on here?

**Comment by JohnBetaro** [comment]
is the development still active on this one ?

**Comment by AudricV** [comment]
Yes, it is.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudChannelInfoItemExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudChannelInfoItemExtractor.java:31]: Resolution added in the list (with more resolutions found) and separate addition removed.

File: extractor/src/main/java/org/schabi/newpipe/extractor/stream/StreamInfo.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/stream/StreamInfo.java:183]: Because the default values are now set in the corresponding attributes.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampExtractorHelper.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampExtractorHelper.java:295]: Usage of this collector has been made in my changes instead of the previous method.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/streaminfoitem/BandcampPlaylistStreamInfoItemExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/streaminfoitem/BandcampPlaylistStreamInfoItemExtractor.java:71]: It is never `null`, so I changed the code to return them in the case they are not empty instead.

File: extractor/src/main/java/org/schabi/newpipe/extractor/Image.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/Image.java:32]: Done, should we enforce non-nullability and non-emptiness by throwing an `IllegalArgumentException` in the constructor if the URL is null or empty? I am not sure if that's a good idea or not.

File: extractor/src/main/java/org/schabi/newpipe/extractor/InfoItem.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/InfoItem.java:65]: Done, annotations added on all `Extractor`s, `InfoItem`s, `InfoItemExtractor`s, `InfoItemsCollector`s and `Info`s.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCRecentKioskExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCRecentKioskExtractor.java:42]: Done

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:96]: Done

File: extractor/src/main/java/org/schabi/newpipe/extractor/Image.java
Comment (TobiGr) [reply extractor/src/main/java/org/schabi/newpipe/extractor/Image.java:39]: I don't think so. We don't do it at other places

**Comment by eightyy8** [comment]
What's stopping this from being merged? If for any reason there is some unwanted drawback to implementing this, any kind of response can help clear the water.

**Comment by Stypox** [comment]
Mainly lack of time. And as you can see this will probably be merged in 0.26.0.

**Comment by JohnBetaro** [comment]
do you have any idea when will 0.26.0 be released ? is it the next version ? or there will be other versions before it ?

thank you for your time

**Comment by Stypox** [comment]
A couple of months I guess. We will merge many new features in 0.26.0

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java
Comment (TobiGr) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java:1]: I'd change the indentation here to make reading the file easier. The same applies to YoutubeMusicArtistInfoItemExtractor as well.
```patch
Subject: [PATCH] Fixes
---
Index: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java b/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java
--- a/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java	(revision 88d443162049f15712ef20678bb358476561be51)
+++ b/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java	(date 1679338926868)
@@ -38,7 +38,8 @@
     public List<Image> getThumbnails() throws ParsingException {
         try {
             return getImagesFromThumbnailsArray(
-                    albumOrPlaylistInfoItem.getObject("thumbnail")
+                    albumOrPlaylistInfoItem
+                            .getObject("thumbnail")
                             .getObject("musicThumbnailRenderer")
                             .getObject("thumbnail")
                             .getArray("thumbnails"));
@@ -63,7 +64,8 @@
 
     @Override
     public String getUrl() throws ParsingException {
-        String playlistId = albumOrPlaylistInfoItem.getObject("menu")
+        String playlistId = albumOrPlaylistInfoItem
+                .getObject("menu")
                 .getObject("menuRenderer")
                 .getArray("items")
                 .getObject(4)
@@ -74,7 +76,8 @@
                 .getString("playlistId");
 
         if (isNullOrEmpty(playlistId)) {
-            playlistId = albumOrPlaylistInfoItem.getObject("overlay")
+            playlistId = albumOrPlaylistInfoItem
+                    .getObject("overlay")
                     .getObject("musicItemThumbnailOverlayRenderer")
                     .getObject("content")
                     .getObject("musicPlayButtonRenderer")
```

**Review by TobiGr — COMMENTED** [review]
Good job. I only found very small things :+1:

**Review by FireMasterK — APPROVED** [review]
I think fundamentally, this is a great architecture to do this!

**Comment by JohnBetaro** [comment]
Any idea when will this be implemented ? Also guys can I ask what are you trying to do that made you so busy with changing the code ?

**Comment by FireMasterK** [comment]
> Any idea when will this be implemented ? Also guys can I ask what are you trying to do that made you so busy with changing the code ?

We're primarily blocked by the changes required at the NewPipe repository. Unfortunately, this PR makes breaking changes to the API and we'd require changes at NewPipe before this can be merged.

I don't think there's anyone working on those changes currently, so this PR is likely not going to be merged soon unless someone makes the necessary changes on the app side. This unfortunately also means other projects (Piped, Libretube, etc) have to wait as well.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeMusicAlbumOrPlaylistInfoItemExtractor.java:1]: I don't think I should change something, as this formatting is used in several places in the extractor, and it doesn't make reading harder in my opinion, especially in these cases.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampExtractorHelper.java
Comment (AudricV) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampExtractorHelper.java:66]: Done, I didn't add the image extension to the comments, as images are returned if you use the PNG extension (even if images returned seem to still be in the JPG format).

**Review by TobiGr — APPROVED** [review]

**Comment by TobiGr** [comment]
I rebased. Most of the conflicts were caused by the playlist description PR and were thus easy to solve. Feel free to copy from my branch: https://github.com/TobiGr/NewPipeExtractor/tree/multiple-images-support

**Review by Stypox — APPROVED** [review]
Thank you!

**Pull request merged**
